### PR TITLE
enhancement(vrl): added get and set event_metadata vrl functions

### LINKED ISSUE #10009: Allow the Datadog API key to be set on events in-flight
I have a use-case where I'd like to set the Datadog API key used in the `datadog_logs` sink based on event metadata.

Currently we support storing the API key used to send logs to the `datadog_agent` source that is then propagated through to be used in the `datadog_logs` sink. This allows Vector to act as a sort of pass-thru transformation layer for the Datadog Agent. The API key is stored in [the event metadata](https://github.com/vectordotdev/vector/blob/ed044b35d21e24c009cf327d475157c9652828e4/lib/vector-core/src/event/metadata.rs#L17-L19) and so is not accessible in transforms (like `remap`).

I'd like to use an enrichment table to look up the API key to use based on event metadata and configure Vector to use this API key in the sink.

A couple of ideas:

1. Move the API key storage/lookup out of event metadata to a field directly on the event.
Pros: straight-forward to manipulate.
Cons: users are more likely to accidentally expose the API key if sending the event to a sink other than the Datadog Logs sink. `store_api_key` on the `datadog_agent` source defaults to `true`.
2. Add a VRL method for setting (and maybe reading?) event metadata. This could look like:
```
set_event_metadata("datadog_api_key", "ABCDEFG")
```

Or a specific method just to set the Datadog API key metadata like:

```
set_datadog_api_key("ABCDEF")
```

Pros: API key is kept hidden and the risk of exposing it is lower.

Cons: this method is a bit odd in the context of VRL in that it directly modifies the event without taking the value to modify as an argument.

I'm currently inclined towards option 2 here but am curious to get thoughts.

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #10009 

This adds `get_event_metadata` and `set_event_metadata` vrl functions.

Currently the only supported metadata field is `datadog_api_key`. I imagine this will expand in future.

There are no tests in the VRL functions since the test framework doesn't support metadata. There is a behaviour test to cover this.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: website/cue/reference/remap/functions/set_event_metadata.cue [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		{
+			name:        "key"
+			description: "The name of the key to set in the metadata."
+			required:    true
+			enum: {
+				"datadog_api_key": "The Datadog api key."
+			}
+			type: ["string"]
+		},
+                {
```
Comment (lucperkins) [inline website/cue/reference/remap/functions/set_event_metadata.cue:19]: lolsob

File: website/cue/reference/remap/functions/set_event_metadata.cue [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		Sets the given field in the event metadata to the provided value.
+		"""
+
+	arguments: [
+		{
+			name:        "key"
+			description: "The name of the key to set in the metadata."
+			required:    true
+			enum: {
+				"datadog_api_key": "The Datadog api key."
```
Comment (lucperkins) [inline website/cue/reference/remap/functions/set_event_metadata.cue:15]: s/api/API

File: tests/behavior/transforms/remap.toml [hunk 3]
```diff
@@ -2255,3 +2255,23 @@ schema.id = "integer"
...
+  source = '''
+      set_event_metadata("datadog_api_key", "iamanapikey")
+      .key = get_event_metadata("datadog_api_key")
+  '''
+[[tests]]
+  name = "remap_metadata"
+  [tests.input]
+    insert_at = "remap_metadata"
+    type = "log"
+    [tests.input.log_fields]
```
Comment (lucperkins) [inline tests/behavior/transforms/remap.toml:2270]: The intention may be more clear here if you do this instead:

```toml
[tests.input]
log_fields = {}
```

File: website/cue/reference/remap/functions/get_event_metadata.cue [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		Returns the value of the given field from the event metadata.
+		"""
+
+	arguments: [
+		{
+			name:        "key"
+			description: "The name of the key to look up in the metadata."
+			required:    true
+			enum: {
+				"datadog_api_key": "The Datadog api key."
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/get_event_metadata.cue:15]: Could we add a note here about this existing if `store_api_key: true` in the `datadog_agent` source?

File: website/cue/reference/remap/functions/set_event_metadata.cue (hunk 2 above)
Comment (jszwedko) [inline website/cue/reference/remap/functions/set_event_metadata.cue:15]: Could we add a note here that this will be used by the `datadog_*` sinks as the API key to send the events with?

**Review by jszwedko — COMMENTED** [review]
I left a couple of doc comments, but this looks good! I'm wondering if we actually need a way to completely remove the API key too. We could use `""` to indicate this, but the `datadog_*` sinks would need to be updated. Since we are in here anyway, could we just add a `remove_event_metadata` function?

As I write this, I wonder if we should suffix field here like `get_metadata_field` to make it clearer that you are just fetching a piece of metadata rather than all of it?

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Given that this function is tied to a Vector-specific implementation detail, I think we should move this function out of stdlib, and into a separate crate (doesn't have to be a new crate, but it _could_ be a "vector-functions" crate if we anticipate more of these), similar to what we did with enrichment table functions.

Alternatively, we remove the hard-coded `datadog_api_key` key in here, and instead allow for VRL to ask for a list of metadata keys through the `Target` trait.

File: website/cue/reference/remap/functions/get_metadata_field.cue [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		{
+			name: "key"
+			description: '''
+				   The name of the field to look up in the metadata.
+				'''
+			required: true
+			enum: {
+				"datadog_api_key": '''
+					   The Datadog API key.
+					   This exists if the `store_api_key` setting is true in the `datadog_agent` source.
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/get_metadata_field.cue:19]:
```suggestion
					   The Datadog API key.
					   
					   This exists if the `store_api_key` setting is true in the `datadog_agent` source.
```

I think the extra line break is necessary if you actually want a line break in the output since wrapped text is rendered as one flowing "line".

File: website/cue/reference/remap/functions/remove_metadata_field.cue [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		{
+			name: "key"
+			description: '''
+				   The name of the field to look up in the metadata.
+				'''
+			required: true
+			enum: {
+				"datadog_api_key": '''
+					   The Datadog API key.
+					   This field will be used by the  `datadog_*` sinks as the API key to send the events with.
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/remove_metadata_field.cue:19]:
```suggestion
					   The Datadog API key.
					   
					   This field will be used by the  `datadog_*` sinks as the API key to send the events with.
```

File: website/cue/reference/remap/functions/set_metadata_field.cue [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+	arguments: [
+		{
+			name:        "key"
+			description: "The name of the field to set in the metadata."
+			required:    true
+			enum: {
+				"datadog_api_key": '''
+					    The Datadog API key.
+					    This field will be used by the  `datadog_*` sinks as the API key to send the events with.
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/set_metadata_field.cue:17]:
```suggestion
					    The Datadog API key.
					    
					    This field will be used by the  `datadog_*` sinks as the API key to send the events with.
```

File: lib/vrl/core/src/target.rs [hunk 8]
```diff
@@ -51,4 +51,12 @@ pub trait Target: std::fmt::Debug {
     /// If `compact` is true, after deletion, if an empty object or array is
     /// left behind, it should be removed as well, cascading up to the root.
     fn remove(&mut self, path: &LookupBuf, compact: bool) -> Result<Option<Value>, String>;
+
+    fn get_metadata(&self, _key: &str) -> Result<Option<vrl_core::Value>, String> {
+        Err("metadata not available".to_string())
+    }
+
+    fn set_metadata(&mut self, _key: &str, _value: String) -> Result<(), String> {
```
Comment (jszwedko) [inline lib/vrl/core/src/target.rs:59]: Do we need `remove_metadata` here too?

Comment (StephenWakely) [reply lib/vrl/core/src/target.rs:59]: That target file is not actually used. I'll delete it.

File: website/cue/reference/remap/functions/get_metadata_field.cue [hunk 9]
```diff
@@ -0,0 +1,39 @@
+package metadata
+
+remap: functions: get_metadata_field: {
+	category: "System"
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/get_metadata_field.cue:4]: Maybe we need a new category, like `Event`? Or `Metadata`?

File: website/cue/reference/remap/functions.cue [hunk 10]
```diff
@@ -26,7 +26,7 @@ remap: {
 		examples?: [remap.#Example, ...remap.#Example]
 	}
 
-	#FunctionCategory: "Array" | "Codec" | "Coerce" | "Convert" | "Debug" | "Enrichment" | "Enumerate" | "Path" | "Hash" | "IP" | "Number" | "Object" | "Parse" | "Random" | "String" | "System" | "Timestamp" | "Type"
+	#FunctionCategory: "Array" | "Codec" | "Coerce" | "Convert" | "Debug" | "Enrichment" | "Enumerate" | "Event" | "Path" | "Hash" | "IP" | "Number" | "Object" | "Parse" | "Random" | "String" | "System" | "Timestamp" | "Type"
```
Comment (lucperkins) [inline website/cue/reference/remap/functions.cue:29]: Nit: could you add this to the `function_categories` array right down below as well?

**Review by JeanMertz — APPROVED** [review]

**Comment by binarylogic** [comment]
@StephenWakely before I open an issue, why did you choose to make each metadata field opt-in? It's very likely we'll forget as we develop this:

https://github.com/vectordotdev/vector/pull/10261#pullrequestreview-824473775

Is there not a way to allow access to all fields? Or use the compiler to force our devs to explicitly handle allowing access or not?

**Comment by StephenWakely** [comment]
> Is there not a way to allow access to all fields?

It's just the way the metadata is represented as fields of a struct which means we have manually maintain a representation for a string -> that field. 

We could definitely make this better, possibly using a macro to wrap the metadata fields that would also generate functions that returned a list of available fields as well as getter and setter functions.

**Comment by JeanMertz** [comment]
To expand on this:

We did discuss this during a check-in. It was deemed this was the easiest way to get this in, and that the team didn't want to invest too much time on the underlying infrastructure for this, as @jszwedko had some doubts on how event metadata would evolve in the future, and whether more (free-form) metadata fields would be added.

In https://github.com/vectordotdev/vector/issues/10009#issuecomment-967002476 I proposed adding specific syntax to manipulate event metadata, which we all agreed is the right direction, but until event metadata is a bit more established, the thought was to go with something simpler, albeit less flexible and (as mentioned) with some manual work needed to keep it up-to-date.

In short, we believe these functions are temporary, until event metadata becomes more of a "first-class citizen" and we can incorporate it into VRL in a more streamlined way.
